### ISSUE TITLE
Could NOT find CUDNN (missing: CUDNN_LIBRARY CUDNN_INCLUDE_DIR)

### ISSUE THREAD

**Issue by brian208579** [opencv/opencv#16380]
OS: win10 ( 64bits )
Cuda : 10.2
Cudnn : 7.6.5
Opencv: github latest

When I use cmake-gui I get this information :

**Could NOT find CUDNN (missing: CUDNN_LIBRARY CUDNN_INCLUDE_DIR) (Required is at least version "7.5")**
![cudnn_problem_1](https://user-images.githubusercontent.com/43407888/72668482-583fbe80-3a62-11ea-89ec-a14ea47f8723.png)

and

**cuDNN : NO**
![cudnn_problem_2](https://user-images.githubusercontent.com/43407888/72668507-9b019680-3a62-11ea-856a-c3cc928d9fae.png)

But I have installed it in the c: / path
please give me some suggestions , thx ...

**Comment by BadMachine** [comment]
@brian208579 i guess it just because you didn`t install cudnn) 
[cudnn download link](https://developer.nvidia.com/rdp/form/cudnn-download-survey)
Cheers!

**Comment by brian208579** [comment]
@BadMachine 

Thanks for your response

But I already have it installed and put it in c: /

![cudnn](https://user-images.githubusercontent.com/43407888/72677520-35e88800-3ad8-11ea-98bf-3541e481a0d8.png)

I don't know what to do ...

**Comment by olramde** [comment]
.

**Comment by BadMachine** [comment]
@brian208579 
Wrong!
You should place in at the same folder that CUDA installed.
E.g. "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\"

**Comment by brian208579** [comment]
@BadMachine 

Sorry , it doesn't work ...
When I configure I get the same result

![cudnn_1](https://user-images.githubusercontent.com/43407888/72677897-f58b0900-3adb-11ea-91c6-d4db32628a10.png)

**Comment by BadMachine** [comment]
@brian208579 you didn't get me.
Now in CUDNN folder you have include/lib/bin.
At the same time in .../CUDA/v10.2/ u have the same folders.
Now you should paste lib/bin/include from CUDNN to .../CUDA?v10.2/

**Comment by brian208579** [comment]
@BadMachine 
Unfortunately ...  It still doesn’t work

**Comment by BadMachine** [comment]
you can write me teamviewer mail and I will try to help , otherwise I can’t help if I don’t see what is happening

**Comment by brian208579** [comment]
@BadMachine 
I already mail, please check it , thank you.

**Comment by brian208579** [comment]
Problem has been solved !!!
thx for @BadMachine

**Comment by haquocviet** [comment]
> Problem has been solved !!!
> thx for @BadMachine

I got the same issue. 
Could you share the solution you have to me and other people?

**Comment by BadMachine** [comment]
@haquocviet First of all make sure u are using release version OpenCV. Then make sure that cudnn folders with files merged with cuda folders: copy 3 folders from cudnn to cuda folder.

**Comment by haquocviet** [comment]
I am using OpenCV 4.2.0 release.
I checked the path again and found nothing abnormal, result is same as before.
![CDNN-folder](https://user-images.githubusercontent.com/14211332/73417983-49f47b00-435d-11ea-9859-468416d4d296.PNG)

```
General configuration for OpenCV 4.2.0 =====================================
  Version control:               4.2.0

NVIDIA CUDA:                   YES (ver 10.2, CUFFT CUBLAS)
    NVIDIA GPU arch:             75
    NVIDIA PTX archs:
  cuDNN:                         NO
```

**Comment by BadMachine** [comment]
@haquocviet I said cudnn folders, not cudnn folder as is. If u don’t understand “merge” meaning:
Copy content from cudnn include dir to cuda include dir, same for lib dir, etc.

**Comment by haquocviet** [comment]
Thanks for your explanation.
That way of course will work but it seems not relate to the issue 
`Could NOT find CUDNN (missing: CUDNN_LIBRARY CUDNN_INCLUDE_DIR) (Required is at least version "7.5")`
CUDNN_LIBRARY and CUDNN_INCLUDE_DIR should point to folders of cuDNN somewhere separately instead in same place as CUDA's.

**Comment by BadMachine** [comment]
@haquocviet ok, gimme treamviewer -_-

**Comment by haquocviet** [comment]
Thanks.
I solved the issue. No need to merge cudnn's files into CUDA folders.
I modified CMakeCache.txt directly to change vars below.
```
CUDNN_INCLUDE_DIR:FILEPATH=D:/.../cudnn-7.5.6/include
CUDNN_LIBRARY:FILEPATH=D:/.../cudnn-7.5.6/lib/cudnn.lib
```

**Comment by solarflarefx** [comment]
@haquocviet how did you go about setting this?  I downloaded cudnn and put the files in the proper directories.  I even ensured the paths are in the environment variables.  However, I still get an error saying cudnn was not detected.  I tried following the instructions on here: https://jamesbowley.co.uk/accelerate-opencv-4-2-0-build-with-cuda-and-python-bindings/

**Comment by haquocviet** [comment]
First, you need to generate CMakeCache.txt file via either cmake cmdline or gui. Then you replace the vars in the file with your real paths of DNN. Finally you do "configure" again to complete the job.
Note that, this is only a temporary way to workaround the issue which the next version of opencv should fix it.

**Comment by cudawarped** [comment]
> @haquocviet how did you go about setting this? I downloaded cudnn and put the files in the proper directories. I even ensured the paths are in the environment variables. However, I still get an error saying cudnn was not detected. I tried following the instructions on here: https://jamesbowley.co.uk/accelerate-opencv-4-2-0-build-with-cuda-and-python-bindings/

Just to confirm when you type the following into cmd

> where cudnn64_7.dll

you get the below path?

> C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\cudnn64_7.dll

**Comment by solarflarefx** [comment]
Thanks for the reply.  I should have mentioned that I am using CUDA 10.1.  If I type "where cudnn64_7.dll" in command line, I get:
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\bin\cudnn64_7.dll
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0\bin\cudnn64_7.dll
C:\libtorch-win-shared-with-deps-debug-1.4.0\libtorch\lib\cudnn64_7.dll

Also I wanted to  confirm another thing -- after doing the build I should be able to reference the .dll and .lib files and use them in a Visual C++ project?  I tried to do so but when I used a GpuMat structure, I got an error saying that OpenCV was not built with CUDA, which is a bit confusing.

I looked at the existing CMakeCache.txt file, which says:
//Include NVidia Cuda Runtime support
WITH_CUDA:BOOL=ON

//List of extra modules
OPENCV_MODULES_EXTRA:INTERNAL=aruco;bgsegm;bioinspired;ccalib;cnn_3dobj;cudaarithm;cudabgsegm;cudacodec;cudafeatures2d;cudafilters;cudaimgproc;cudalegacy;cudaobjdetect;cudaoptflow;cudastereo;cudawarping;cudev;cvv;datasets;dnn_objdetect;dnn_superres;dpm;face;freetype;fuzzy;hdf;hfs;img_hash;intensity_transform;java_bindings_generator;line_descriptor;matlab;optflow;ovis;phase_unwrapping;plot;python2;python3;python_bindings_generator;python_tests;quality;rapid;reg;rgbd;saliency;sfm;shape;stereo;structured_light;superres;surface_matching;text;tracking;videostab;viz;xfeatures2d;ximgproc;xobjdetect;xphoto
//List of main modules
OPENCV_MODULES_MAIN:INTERNAL=core;imgproc;imgcodecs;videoio;highgui;video;calib3d;features2d;objdetect;dnn;ml;flann;photo;stitching;gapi;java;js;ts;world
//List of OpenCV modules marked for export
OPENCV_MODULES_PUBLIC:INTERNAL=opencv_calib3d;opencv_core;opencv_dnn;opencv_features2d;opencv_flann;opencv_highgui;opencv_imgcodecs;opencv_imgproc;opencv_ml;opencv_objdetect;opencv_photo;opencv_stitching;opencv_video;opencv_videoio;opencv_world;opencv_aruco;opencv_bgsegm;opencv_bioinspired;opencv_ccalib;opencv_cudaarithm;opencv_cudabgsegm;opencv_cudacodec;opencv_cudafeatures2d;opencv_cudafilters;opencv_cudaimgproc;opencv_cudalegacy;opencv_cudaobjdetect;opencv_cudaoptflow;opencv_cudastereo;opencv_cudawarping;opencv_cudev;opencv_datasets;opencv_dnn_objdetect;opencv_dnn_superres;opencv_dpm;opencv_face;opencv_fuzzy;opencv_hfs;opencv_img_hash;opencv_intensity_transform;opencv_line_descriptor;opencv_optflow;opencv_phase_unwrapping;opencv_plot;opencv_quality;opencv_rapid;opencv_reg;opencv_rgbd;opencv_saliency;opencv_shape;opencv_stereo;opencv_structured_light;opencv_superres;opencv_surface_matching;opencv_text;opencv_tracking;opencv_videostab;opencv_xfeatures2d;opencv_ximgproc;opencv_xobjdetect;opencv_xphoto

It seems that the build was set up to build with CUDA but my results show otherwise.

**Comment by cudawarped** [comment]
Can you post the output to the below please

> C:\opencv_build\opencv\build\bin\opencv_version_win32d.exe

or if you built release

> C:\opencv_build\opencv\build\bin\opencv_version_win32.exe

What size is your shared library?

> C:\opencv_build\opencv\build\bin\opencv_world420d.dll

Do you have either

> C:\opencv_build\opencv\build\bin\opencv_perf_cudaarithmd.exe

or 

> C:\opencv_build\opencv\build\bin\opencv_perf_cudaarithm.exe

if so do some of the tests pass?

**Comment by solarflarefx** [comment]
Regarding running opencv_version_win32d.exe:
![Q1_1](https://user-images.githubusercontent.com/45402633/75297828-da0fce00-57fe-11ea-8b89-1f3fdf89ebc8.PNG)
![Q1_2](https://user-images.githubusercontent.com/45402633/75297829-daa86480-57fe-11ea-9d30-2b2b0d16c33a.PNG)
![Q1_3](https://user-images.githubusercontent.com/45402633/75297830-daa86480-57fe-11ea-9fc2-c46c433549e8.PNG)
![Q1_4](https://user-images.githubusercontent.com/45402633/75297832-db40fb00-57fe-11ea-9c39-bccbd8b80dfb.PNG)

Size of "Debug" folder within C:\opencv_build\opencv\build\bin is: 6.40 GB

I do have C:\opencv_build\opencv\build\bin\Debug\cudaarithmd.exe
Results:
![Cuda_Arithmd_1](https://user-images.githubusercontent.com/45402633/75298217-e6485b00-57ff-11ea-8a5e-538148a74e8e.PNG)
![Cuda_Arithmd_2](https://user-images.githubusercontent.com/45402633/75298218-e6e0f180-57ff-11ea-82c6-64e32394e38c.PNG)
![Cuda_Arithmd_3](https://user-images.githubusercontent.com/45402633/75298219-e6e0f180-57ff-11ea-8802-6b8564e095e2.PNG)

**Comment by alalek** [comment]
Avoid posting screenshots with text information - it is ridiculous.

> CMakeCache.txt

Check contents of CMakeVars.txt for CUDA / CUDNN stuff too.

If you make experiments with dependencies, then you should clean **CMake cache** before each run.

(*May not work*) Partial cache cleanup is possible:
- `cmake . -U*CUDA* -U*CUDNN* ...-Dnew=values...`

**Comment by cudawarped** [comment]
The output implies you have managed to finish the build of OpenCV with CUDA support, my guess would by as alalek has suggested, you probably ran cmake initialy before you had placed cudnn64_7.dll on your path (or you placed it on your system path without a restart or your user path without opening a new command prompt or similar).

Regarding the error you mentioned in another thread 

> Perhaps I do not need cudnn for the GpuMat use, but when I used the .dll and .lib files, Visual Studio gave an error saying they were not built with CUDA support 

I can only think that you are not picking up the correct .dll at runtime as the results of running

> C:\opencv_build\opencv\build\bin\opencv_perf_cudaarithmd.exe

show that you can successfully build and run an application against 

> C:\opencv_build\opencv\build\bin\opencv_world420d.dll

compiled with CUDA support.

**Comment by solarflarefx** [comment]
Thanks for the help and guidance.

I apologize for the screenshots on the text information -- I will copy and paste text for this kind of information.

So I looked at my configuration for Visual Studio and it is the following:

VC++ Directories
----------------------------
Include Directories: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include;C:\opencv_build\opencv\build\install\include;$(IncludePath)

Library Directories: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64;C:\opencv_build\opencv\build\lib\Debug;$(LibraryPath)
----------------------------

C/C++
----------------------------
Additional Include Directories: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include
----------------------------

Linker:
----------------------------
Additional library directories: C:\opencv_build\opencv\build\lib\Debug;
----------------------------

Additional Dependencies:
----------------------------
opencv_world420d.lib;%(AdditionalDependencies)
----------------------------


My code is pretty simple:
`
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <iostream>
#include <opencv2/cudaarithm.hpp>

using namespace cv;
using namespace std;

int main()
{
	Mat image = Mat::zeros(300, 600, CV_8UC1);
	cv::cuda::GpuMat gpu_image(image);
	circle(image, Point(250, 150), 100, Scalar(0, 255, 128), -100);
	circle(image, Point(350, 150), 100, Scalar(255, 255, 255), -100);
	imshow("Display Window", image);
	waitKey(0);
	return 0;
}
`

I get the following error:
OpenCV(4.2.0) Error: No CUDA support (The library is compiled without CUDA support) in throw_no_cuda, file C:\build\master_winpack-build-win64-vc15\opencv\modules\core\include\opencv2/core/private.cuda.hpp, line 109


My guess is I am referencing the wrong files in the configuration?  My point of confusion is that I see multiple files of the same name on the C:\opencv_build directory and am not 100% sure which ones are the "correct" ones.

**Comment by cudawarped** [comment]
> Regarding not having restarted or opened a new command prompt, I had actually done both and it seems that cudnn still does not show up.

Did you as alalek suggested clean the cache or even better delete or use a fresh build directory?

> OpenCV(4.2.0) Error: No CUDA support (The library is compiled without CUDA support) in throw_no_cuda, file C:\build\master_winpack-build-win64-vc15\opencv\modules\core\include\opencv2/core/private.cuda.hpp, line 109

I assume you are seeing this error when running your program?

If so check the location reported by

> where opencv_world420d.dll

is only

> C:\opencv_build\opencv\build\bin\Debug\opencv_world420d.dll

If you are using visual studio you should be able to see this location in the modules window when debugging your program.

Additionally if you have built with Visual Studio, you have a multiple working projects you can study (opencv_test_cudaarithm etc.) when the solution is open to get build options which work.

**Comment by solarflarefx** [comment]
Hello,

So I took a look at some of the Visual Studio solution files and it seems that whenever I try to build and compile them, it seems to be building everything from source again to run an executable.

Perhaps my understanding is wrong, but one of my goals was to build OpenCV with CUDA and then use the .dll and .lib files to create my own Visual Studio C++ project.  I did not see any of the Visual Studio solution examples doing that.  As I mentioned, when I tried to include the .dll and .lib files and then tried to run my code in Debug mode, I got an error saying that OpenCV was not built with CUDA.

Is my understanding and approach incorrect?  Thanks for all the help thus far.

**Comment by cudawarped** [comment]
You can select this project only and then build to avoid building the whole solution.
Please check which .dll is loaded in the module windows when debugging and/or by using the where command I mentioned above , or just copy 

> C:\opencv_build\opencv\build\bin\opencv_world420d.dll

to your build directory to force the correct .dll to be used.

**Comment by solarflarefx** [comment]
So I notice I do not have any .lib or .dll files in the directory: C:\opencv_build\opencv\build\bin

I have a "Debug" and "Release" directory in there though.  In the "Debug" directory I have .pdb, .ilk, and .exe files.  No .dll or .lib files however.  Am I in the wrong directory, or did something go wrong in my build process?

**Comment by manan-d8** [comment]
If Anyone Doing this on Windows try deleting cache from build folder in cmake or change build folder...!

**Comment by Adriann2k** [comment]
Anyone being able to help me with this? @BadMachine or @haquocviet 

I've been trying 10 hours to install this, and i am still failing. I am really desperate. Please, help me.

Thanks!

**Comment by cudawarped** [comment]
@Adriann2k 
What is your exact issue?  Is it just that you are not picking up cuDnn when running cmake or that you are failing to compile OpenCv with CUDA, can you post your errors?

**Comment by Adriann2k** [comment]
i managed to fix it by downgrading python from 3.7 to 3.6

**Comment by Adriann2k** [comment]
thanks anyways :)

**Comment by HduSy** [comment]
> 
> 
> Problem has been solved !!!
> thx for @BadMachine

I still haven't solved this problem,i copy the include\lib to the target dir:usr/local/cuda/include and lib,but it didn't work
my cuda is 10.0 cudnn is v7.5 for cuda10.0

**Comment by HduSy** [comment]
Actually you can also solve by cmake add this
`-D CUDNN_INCLUDE_DIR="your cudnn include path" 
-D CUDNN_LIBRARY="your cudnn lib path" `

**Comment by bigntallmike** [comment]
> Actually you can also solve by cmake add this
> `-D CUDNN_INCLUDE_DIR="your cudnn include path" -D CUDNN_LIBRARY="your cudnn lib path" `

This is what I did to solve it as well.  In my case:
```
    -D CUDNN_INCLUDE_DIR=/usr/include \
    -D CUDNN_LIBRARY=/usr/lib64/libcudnn_static_v7.a \
    -D CUDNN_VERSION=7.6.5 \
```

**Comment by nhorton04** [comment]
`-D CUDNN_INCLUDE_DIR="your cudnn include path" -D CUDNN_LIBRARY="your cudnn lib path"`
I tried adding these to cmake-gui and it still didn't see CUDNN :'(

**Comment by taf2** [comment]
@nhorton04 

I got this working for me by doing the following:

```
sudo apt install libcudnn7-dev  libcudnn7-doc  libcudnn7 nvidia-container-csv-cudnn
```
Checked to see which version of cudnn it installed, for me it's 7.6.3

```
 -D CUDNN_INCLUDE_DIR=/usr/include \
-D CUDNN_LIBRARY=/usr/lib64/libcudnn_static_v7.a \
-D CUDNN_VERSION=7.6.3
```

hope this helps!

**Comment by rickymedrano** [comment]
@nhorton04 
I was getting the same errors as you. I fixed it by including this in the cmake command:
	-D CUDNN_INCLUDE_DIR=/usr/include \
	-D CUDNN_LIBRARY=/usr/local/cuda-10.2/lib64/libcudnn.so.7 \
	-D CUDNN_VERSION=7.6.5 \
Note your libcuddn.so.7 file could be in cuda, cuda-10.1 or cuda-10.2 so adjust the path accordingly above.

**Comment by NuwanMadhusanka0417** [comment]
cmake -D CMAKE_BUILD_TYPE=RELEASE -D CMAKE_INSTALL_PREFIX=/usr/local -D INSTALL_PYTHON_EXAMPLES=ON -D INSTALL_C_EXAMPLES=OFF -D OPENCV_ENABLE_NONFREE=ON -D WITH_CUDA=ON -D WITH_CUDNN=ON -D OPENCV_DNN_CUDA=ON -D ENABLE_FAST_MATH=1 -D CUDA_FAST_MATH=1 -D CUDA_ARCH_BIN=6.1 -D WITH_CUBLAS=1 -D OPENCV_EXTRA_MODULES_PATH=~/opencv_contrib/modules -D HAVE_opencv_python3=ON -D PYTHON_EXECUTABLE=~/env/bin/python -D BUILD_EXAMPLES=ON -D CUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda-11.0 -D CUDNN_INCLUDE_DIR=/usr/include -DCUDNN_LIBRARY=/usr/local/cuda/lib64/libcudnn.so.8 -D CUDNN_VERSION=8.0.3.33 ..


This is work for me.. Thank you all

**Comment by SkpC9** [comment]
I am using windows with cuda10.1 and cudnn7.6.5.
For anyone who gets the same error, I added `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include` `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\Win32` `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64` to System PATH, then Cmake can find CUDNN.

**Comment by asimukaye** [comment]
Hi, I was facing a similar issue and none of the above fixes worked for me. I was using OpenCV 4.2.0 with CUDA 10.2 and CUDNN 8.2.
After several attempts, I figured that the FindCUDNN.cmake in OpenCV cmake folder was looking for the version of CUDNN in "cudnn.h" whereas CUDNN 8.2 has the version in "cudnn_version.h". So I replaced the name in FindCUDNN.cmake and it worked finally.

P.S : clear all the cmake cache before trying again

**Comment by tellamon** [comment]
I had this problem too.

I am using windows with cuda10.1, cudnn8.0.5.39, cuda10.1

My solution : 
1. I merged cudnn files to CUDA folder.
2. I removed the unnecessary PATH. (no need to add CUDNN_LIBRARY or CUDNN_INCLUDE_DIR) and also Anaconda path. (I found that anaconda had cudnn dll with different version)
3. Reboot.
4. I just cleaned up CMake build folder and re-configure several times with CUDA option enabled.
5. cmake configure works! ( cuDNN : YES (ver 8.0.5))
